liteeth/mac/sram.py

```diff
@@ -62,23 +68,58 @@ def __init__(self, dw, depth, nslots=2, endianness="big"):
         ongoing = Signal()
 
         # Status FIFO
-        fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
+        if timestamp_source is not None:
+            fifo = stream.SyncFIFO([
+                ("slot", slotbits),
+                ("length", lengthbits),
+                ("rx_timestamp", timestampbits),
+            ], nslots)
+        else:
+            fifo = stream.SyncFIFO([("slot", slotbits), ("length", lengthbits)], nslots)
         self.submodules += fifo
 
+        # If we timestamp incoming packets we're going to need another
+        # signal to hold the timestamp from the start of packet
+        # reception
+        if timestamp_source is not None:
+            rx_start_timestamp = Signal(timestampbits)
+
+            # Effectively build a latch, updating the timestamp only
+            # when starting the reception
+            self.comb += [
+                rx_start_timestamp.eq(rx_start_timestamp)
+            ]
```

Comment:
You can use single line` self.comb +=  rx_start_timestamp.eq(rx_start_timestamp)` for this.

Reply:
In fact, we should avoid latches, so you can probably remove this code and use `NextValue(rx_start_timestamp, timestamp_source)` in the FSM.

Reply:
Ah, this is cool. I'm still very new to Migen and haven't used Python a lot, so I wanted to "just make it work". :)